File: leetcode2904/Solution.go
// Solution using enumeration

func shortestBeautifulSubstring(s string, k int) string {
	for m := k; m <= len(s); m++ {
		ans := ""
		for i := m; i <= len(s); i++ {
			t := s[i-m : i]
			if (ans == "" || t < ans) && strings.Count(t, "1") == k {
				ans = t
			}
		}
		if ans != "" {
			return ans
		}
	}
	return ""
}

// Solution using sliding window
func shortestBeautifulSubstring(s string, k int) string {
	if strings.Count(s, "1") < k {
		return ""
	}
	ans := s
	cnt, left := 0, 0
	for right := 0; right < len(s); right++ {
		cnt += int(s[right] - '0')
		for cnt > k || s[left] == '0' {
			cnt -= int(s[left] - '0')
			left++
		}
		if cnt == k {
			t := s[left : right+1]
			if len(t) < len(ans) || len(t) == len(ans) && t < ans {
				ans = t
			}
		}
	}
	return ans
}

File: leetcode2904/Solution.cpp
using namespace std;
/**
 * ### SLIDING WINDOW SOLUTION ###
 */
class Solution {
public:
     string shortestBeautifulSubstring(string s, int k) {
          if (ranges::count(s, '1') < k) {
               return "";
          }
          string ans = s;
          int cnt = 0;
          for (int left = 0, right = 0; right < s.length(); right++) {
               cnt += s[right] - '0';
               while (cnt > k || s[left] == '0') {
                    cnt -= s[left] - '0';
                    left++;
               }
               if (cnt == k) {
                    string t = s.substr(left, right - left + 1);
                    // check for smallest length wise and lexicographically
                    if (t.length() < ans.length() ||
                         t.length() == ans.length() && t < ans) {
                         ans = move(t);
                    }
               }
          }
          return ans;
     }
};

/**
 * ### SOLUTION USING ENUMERATION ###
 * class Solution {
 * public:
 *   string shortestBeautifulSubstring(string s, int k) {
 *        for (int m = k; m <= s.length(); m++) {
 *             string ans = "";
 *             for (int i = m; i <= s.length(); i++) {
 *                  string t = s.substr(i - m, m);
 *                  if ((ans.empty() || t < ans) && ranges::count(t, '1') == k) {
 *                       ans = t;
 *                  }
 *             }
 *             if (!ans.empty()) {
 *                  return ans;
 *             }
 *        }
 *        return "";
 *   }
 * };
 */